src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/EditorConfigFileWatcher.cs

```diff
@@ -0,0 +1,271 @@
+                    await fileChangeService.AdviseFileChangeAsync(
+                        editorConfigFilePath,
+                        _VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del,
+                        sink: this);
+            }
+        }
+
+        /// <summary>
+        /// Initialize the watcher.
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Explicitly get back to the thread pool for the rest of this method so we don't tie up the UI thread;
+            await TaskScheduler.Default;
+
+            await _projectTasksService.LoadedProjectAsync(() =>
+                {
+                    _filesBeingWatched = new Dictionary<string, uint>(StringComparers.Paths);
+                    _taskDelaySchedulerCancellationTokenSource = CreateLinkedCancellationTokenSource();
+                    _taskDelayScheduler = new TaskDelayScheduler(
+                        s_notifyDelay,
+                        _projectServices.ThreadingService,
+                        _taskDelaySchedulerCancellationTokenSource.Token);
+
+                    // The tree source to get changes to the tree so that we can identify *when* the set of .editorconfig files changes.
+                    ProjectDataSources.SourceBlockAndLink<IProjectVersionedValue<IProjectTreeSnapshot>> treeSource = _fileSystemTreeProvider.Tree.SyncLinkOptions();
+
+                    // The source to get the actual set of .editorconfig files
+                    StandardRuleDataflowLinkOptions sourceLinkOptions = DataflowOption.WithRuleNames(ConfigurationGeneral.SchemaName);
+
+                    ProjectDataSources.SourceBlockAndLink<IProjectVersionedValue<IProjectSubscriptionUpdate>> editorconfigItemsSource = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock.SyncLinkOptions(sourceLinkOptions);
+                    ITargetBlock<IProjectVersionedValue<Tuple<IProjectTreeSnapshot, IProjectSubscriptionUpdate>>> target = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<Tuple<IProjectTreeSnapshot, IProjectSubscriptionUpdate>>>(DataFlow_ChangedAsync);
+
+                    // Join the two sources so that we get synchronized versions of the data.
+                    _treeWatcher = ProjectDataSources.SyncLinkTo(treeSource, editorconfigItemsSource, target);
+
+                    return Task.CompletedTask;
+                });
+        }
+
+        private CancellationTokenSource CreateLinkedCancellationTokenSource()
+        {
+            // we want to cancel when we switch what file is watched, or when the project is unloaded
+            if (_projectServices.Project?.Services?.ProjectAsynchronousTasks?.UnloadCancellationToken != null)
+            {
+                return CancellationTokenSource.CreateLinkedTokenSource(
+                    _projectServices.Project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
```

Comment:
Question: Does this need to link to this classes Dispose token? Or would it do the same job as this effectively?

Reply:
Yes, it should just link both DisposalToken/UnloadCancellationToken. I see you copied this code from ProjectFileWatcher but it's incorrect, down to the null checks.